rfcs/0028-nix-release-model.md

```diff
@@ -0,0 +1,158 @@
+---
+feature: nix-release-model
+start-date: 2018-03-28
+author: Shea Levy <shea@shealevy.com>
+co-authors: (find a buddy later to help our with the RFC)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+A model of when, what, and how to make new releases of Nix.
+
+# Motivation
+[motivation]: #motivation
+
+We want to set clear expectations for users, provide guidance for
+developers, and provide both groups the benefits of easy availability
+of new improvements.
+
+# Detailed design
+[design]: #detailed-design
+
+The proposed process contains a number of interrelated components.
+I've numbered them for ease of reference and discussion, not to
+indicate any meaningful ordering. Any aspect of this can be suspended
+at the security team's discretion when handling a vulnerability. Items
+marked with a * need technical work to implement, which needn't delay
+the rest of the process.
+
+1. Keep `master` as close to release readiness as possible. This
+   includes testing, documentation, changelog entries, etc. being
+   ready *before* merge.
+2. \*Use code review and automated checks, enforced through branch
+    protection, to ensure point 1 (and generally improve quality and
+    increase knowledge spread).
+3. Follow semver. In light of 1, this includes ensuring any necessary
+   semver bumps happen before a relevant change is merged.
+4. Any community member may request a new release by opening an issue
+   on GitHub.
+5. If any Nix core team member thinks the release is warranted, they
+   can assign themselves the ticket and assume release manager
+   responsibilities for the release, barring another team member's
+   explained disagreement with the timing. Core team members are
+   trusted to use their best judgment in this decision, balancing both
+   the needs of the community at large and the needs of developers
+   working on Nix. In particular, if an especially major change has
+   landed recently or is nearing readiness, the release manager should
+   either branch off before the relevant change or be prepared to do
+   extra stabilization work.
+6. If no core team member is available to do the release work but the
+   team agrees (through its normal decision procedures) that a release
+   is warranted, a volunteer from the community can serve as release
+   manager.
+7. Under normal circumstances, there will be at least one mainline
+   release between each `NixOS` stable release. An explicit decision
+   *not* to release should be made by the core team if a stable
+   release is approaching and no new release has been made.
+8. The release manager will branch off a `#-stabilization` branch
+   (e.g. `2.0.1-stabilization`) in advance of any release.
```

Comment:
if master is always in a releasable state (according to point 1), is this branch necessary?

Reply:
Yes, because a «releasable state» is not a completely formal notion. Stabilisation process can attract wider testing (and wider reading of documentation) to make sure nothing has been missed.

Reply:
Yeah, exactly. If nothing else, letting things cook in `nixUnstable` for a while is essential IMO.

Reply:
To make it easier to release bug-fixes it might make sense to create `release-2.0` branch that contains all the tiny releases.

Master can keep evolving with major features without having to worry on impairing release.

If a bug fix release is needed, cherry-pick the fixes from master onto the `release-2.0` branch until it's "ready" and then tag the release from there.

If a new major or minor release is required, branch off from master and create the new `release-MAJOR.MINOR` branch. Test until it's stable and tag the `MAJOR.MINOR.0` release.

This would be closer to the nixpkgs channel releases where master is constantly evolving and patches get cherry-picked into the release channels.

Reply:
I think what you describe is contained in points 15–18 about maintenance branches. But if everything goes smoothly, they are not mandatory.

Reply:
That's what I am saying, stabilization and maintenance branches are essentially the same thing.

Reply:
Well, maintenance branches are used for more than one release. I think minimising the divergence when possible is a worthy goal, so as long as it makes sense to just release the point-release from master branch, it is allowed.

Reply:
In any case, as long as the nix-core team agrees on the process, all my input can be disregarded.

Reply:
Well, we don't know for sure the opinions of most of the Nix core team (only @shlevy), and maybe your comments would be used as a basis of an alternative approach.

Of course I agree that your approach has its benefits, it's all tradeoffs.